python/triqs_dft_tools/sumk_dft.py

```diff
@@ -1981,14 +2091,94 @@ def calc_mu(self, precision=0.01, broadening=None, delta=0.5, max_loops=100):
+                y1=y2
+
+                x2 -=  eps*delta_x
+                y2 = function(x2)
+
+            if nbre_loop > (max_loops):
+                raise ValueError("The bounds could not be found")
+
+            # Make sure that x2 > x1
+            if x1 > x2:
+                x1,x2 = x2,x1
+                y1,y2 = y2,y1
+
+
+            mpi.report(f"mu_interval: [  {x1:.4f}  ; {x2:.4f} ]")
+            mpi.report(f"delta to target density interval: [ {y1:.4f} ; {y2:.4f} ]")
+            return x1, x2
+
+
+
+        # previous implementation
+        def F_bisection(mu): return self.total_density(mu=mu, broadening=broadening).real
         density = self.density_required - self.charge_below
-
-        self.chemical_potential = dichotomy.dichotomy(function=F,
-                                                      x_init=self.chemical_potential, y_value=density,
-                                                      precision_on_y=precision, delta_x=delta, max_loops=max_loops,
-                                                      x_name="Chemical Potential", y_name="Total Density",
-                                                      verbosity=3)[0]
+        #using scipy.optimize
+        def F_optimize(mu):
+
+            mpi.report("Trying out mu = {}".format(str(mu)))
+            calc_dens = self.total_density(mu=mu, broadening=broadening).real - density
+            mpi.report("Delta to target density = {}".format(str(calc_dens)))
+            return calc_dens
+        
+        #check for lowercase matching for the method variable
+        match method.lower():
+
+            case "dichotomy":
+                mpi.report("\nSUMK calc_mu: Using dichtomy adjustment to find chemical potential\n")
+                self.chemical_potential = dichotomy.dichotomy(function=F_bisection,
+                                                              x_init=self.chemical_potential, y_value=density,
+                                                              precision_on_y=precision, delta_x=delta, max_loops=max_loops,
+                                                              x_name="Chemical Potential", y_name="Total Density",
+                                                              verbosity=3)[0]
+            case 'newton':
+                mpi.report("\nSUMK calc_mu: Using Newton method to find chemical potential\n")
+                self.chemical_potential =                newton(func=F_optimize,
+                                                              x0=self.chemical_potential,
+                                                              tol=precision, maxiter=max_loops,
+                                                              )
+
+            case 'brent':
+                mpi.report("\nSUMK calc_mu: Using Brent method to find chemical potential")
+                mpi.report("SUMK calc_mu: Finding bounds \n")
+                
+                mu_guess_0, mu_guess_1  =  find_bounds(function=F_optimize,
```

Comment:
I am not sure how many function evaluations are needed to find the bounds, but we have template of a function to find the bounds without any call to extract_G_loc: https://github.com/TRIQS/lattice_tools_dev/blob/3bfa050f4c806b495fcdbe596d849ce45fadaede/python/dfttools_dev/pytools.py#L428 . Let's discuss.